### ISSUE TITLE
Can't start oficial docker container on arm64

### ISSUE THREAD

**Issue by h2xtreme** [prometheus/prometheus#6943]
**What did you do?**

docker-compose to run the docker container on a arm64 device

**What did you expect to see?**

prometheus up and running

**What did you see instead? Under which circumstances?**

The container crashes

**Environment**

Docker-CE 

* System information:

Ubuntu 20.04

Linux 5.6.0-rc4-arm-64 aarch64

* Prometheus version:

	2.16.0

go version: 1.13.8 on the container
go version: 1.14 on the OS

* Prometheus configuration file:
```
# my global config
global:
  scrape_interval:     15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

# Alertmanager configuration
alerting:
  alertmanagers:
  - static_configs:
    - targets:
      # - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
    - targets: ['localhost:9090']
```

* Logs:
```
level=info ts=2020-03-08T23:43:10.092Z caller=main.go:295 msg="no time or size retention was set so using the default time retention" duration=15d
level=info ts=2020-03-08T23:43:10.093Z caller=main.go:331 msg="Starting Prometheus" version="(version=2.16.0, branch=HEAD, revision=b90be6f32a33c03163d700e1452b54454ddce0ec)"
level=info ts=2020-03-08T23:43:10.093Z caller=main.go:332 build_context="(go=go1.13.8, user=root@7ea0ae865f12, date=20200214-00:59:16)"
level=info ts=2020-03-08T23:43:10.093Z caller=main.go:333 host_details="(Linux 5.6.0-rc4-arm-64 #20.05.0 SMP PREEMPT Tue Mar 3 16:58:21 MSK 2020 aarch64 6253d342969a (none))"
panic: syscall.Getrlimit failed: operation not permitted

goroutine 1 [running]:
github.com/prometheus/prometheus/pkg/runtime.getLimits(0x7, 0x0, 0x0, 0x2, 0x2)
        /app/pkg/runtime/limits_default.go:39 +0x2b4
github.com/prometheus/prometheus/pkg/runtime.FdLimits(...)
        /app/pkg/runtime/limits_default.go:46
main.main()
        /app/cmd/prometheus/main.go:334 +0x3fe4
level=info ts=2020-03-08T23:43:13.384Z caller=main.go:295 msg="no time or size retention was set so using the default time retention" duration=15d
level=info ts=2020-03-08T23:43:13.385Z caller=main.go:331 msg="Starting Prometheus" version="(version=2.16.0, branch=HEAD, revision=b90be6f32a33c03163d700e1452b54454ddce0ec)"
level=info ts=2020-03-08T23:43:13.385Z caller=main.go:332 build_context="(go=go1.13.8, user=root@7ea0ae865f12, date=20200214-00:59:16)"
level=info ts=2020-03-08T23:43:13.386Z caller=main.go:333 host_details="(Linux 5.6.0-rc4-arm-64 #20.05.0 SMP PREEMPT Tue Mar 3 16:58:21 MSK 2020 aarch64 6253d342969a (none))"
panic: syscall.Getrlimit failed: operation not permitted

goroutine 1 [running]:
github.com/prometheus/prometheus/pkg/runtime.getLimits(0x7, 0x0, 0x0, 0x2, 0x2)
        /app/pkg/runtime/limits_default.go:39 +0x2b4
github.com/prometheus/prometheus/pkg/runtime.FdLimits(...)
        /app/pkg/runtime/limits_default.go:46
main.main()
        /app/cmd/prometheus/main.go:334 +0x3fe4
level=info ts=2020-03-08T23:43:16.827Z caller=main.go:295 msg="no time or size retention was set so using the default time retention" duration=15d
level=info ts=2020-03-08T23:43:16.828Z caller=main.go:331 msg="Starting Prometheus" version="(version=2.16.0, branch=HEAD, revision=b90be6f32a33c03163d700e1452b54454ddce0ec)"
level=info ts=2020-03-08T23:43:16.828Z caller=main.go:332 build_context="(go=go1.13.8, user=root@7ea0ae865f12, date=20200214-00:59:16)"
level=info ts=2020-03-08T23:43:16.829Z caller=main.go:333 host_details="(Linux 5.6.0-rc4-arm-64 #20.05.0 SMP PREEMPT Tue Mar 3 16:58:21 MSK 2020 aarch64 6253d342969a (none))"
panic: syscall.Getrlimit failed: operation not permitted

goroutine 1 [running]:
github.com/prometheus/prometheus/pkg/runtime.getLimits(0x7, 0x0, 0x0, 0x2, 0x2)
        /app/pkg/runtime/limits_default.go:39 +0x2b4
github.com/prometheus/prometheus/pkg/runtime.FdLimits(...)
        /app/pkg/runtime/limits_default.go:46
main.main()
        /app/cmd/prometheus/main.go:334 +0x3fe4
```

**Comment by brian-brazil** [comment]
It makes more sense to ask questions like this on the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users) rather than in a GitHub issue. On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.

**Comment by h2xtreme** [comment]
For reference, if anyone has this issue my fix was to add 

privileged: true

in the prometheus section of the docker compose

**Comment by ja0nz** [comment]
Seriously thank you! I had this issue too. While it ran smoothly without privileged on the x86-64 the aarch64 giving me a hard time with those strange errors. Solved!

**Comment by violentmagician** [comment]
I cannot use the `privileged: true` workaround since I am running in swarm mode. Can this bug be reopened since this affects one architecture and not others?

**Comment by roidelapluie** [comment]
It looks as if this is actually a question about usage and not development.

It makes more sense to ask questions like this on the [prometheus-users mailing list](https://groups.google.com/forum/#!forum/prometheus-users) rather than in a GitHub issue. On the mailing list, more people are available to potentially respond to your question, and the whole community can benefit from the answers provided.

**Comment by paulfantom** [comment]
For those who get this error again in the future.

This is not an issue of prometheus container image running on amr64 (many people, including me don't have this problem on ARM or ARM64). This is solely an issue of misconfigured host.

Main problem here are too low hard limits set in system and passed down to container runtime. By default container is not allowed to raise limits over what is enforced system-wide, this can be changed by passing `privileged: true` to container definition as a workaround. However proper solution is to [adjust soft and hard limits on host](http://geekswing.com/geek/quickie-tutorial-ulimit-soft-limits-hard-limits-soft-stack-hard-stack/).

**Comment by h2xtreme** [comment]
@paulfantom doesn't seem like limit config. The container implodes on a function called Getrtlimits, trying to get the host OS limits not setting them.

From the looks of it, its a golang bug in the method used to perform the syscall to get the limits on arm64 arch. More info:

https://github.com/golang/go/issues/38604
https://go-review.googlesource.com/c/go/+/230339/
https://go-review.googlesource.com/c/sys/+/230478/

These were merged on April 28th (I opened this on March 8th), so any docker image of prometheus that uses a golang release after April 28th should have this patch baked in.

On my setup I pulled the latest docker image (uses go1.14.4) and relaunched without priveleged: true and so far its working. No changes performed on limits (Ubuntu defaults, pretty high already)

@violentmagician and @ja0nz you might want to revisit this and pull the latest prometheus docker image and go from there.

**Comment by violentmagician** [comment]
No luck unfortunately.

```
$ docker run --name prometheus -p 9090:9090 prom/prometheus
Unable to find image 'prom/prometheus:latest' locally
latest: Pulling from prom/prometheus
8e1247db4aa5: Already exists
84f1838a5762: Already exists
2064fd4307c7: Pull complete
8a99b247c4ae: Pull complete
6fa9a5d88ba3: Pull complete
9f2d2dc552df: Pull complete
b6e62dd4d487: Pull complete
d390c1ee7a05: Pull complete
cebf6feafbcf: Pull complete
0f9e33c47849: Pull complete
cd1e03f0b87d: Pull complete
3df8a7ce6303: Pull complete
Digest: sha256:efe62fa8804e9fd2612a945b70c630cc27e21b5fb8233ccc8be4cfbe06d26b04
Status: Downloaded newer image for prom/prometheus:latest
level=info ts=2020-06-26T20:31:42.783Z caller=main.go:302 msg="No time or size retention was set so using the default time retention" duration=15d
level=info ts=2020-06-26T20:31:42.783Z caller=main.go:337 msg="Starting Prometheus" version="(version=2.19.1, branch=HEAD, revision=eba3fdcbf0d378b66600281903e3aab515732b39)"
level=info ts=2020-06-26T20:31:42.784Z caller=main.go:338 build_context="(go=go1.14.4, user=root@62700b3d0ef9, date=20200618-17:49:22)"
level=info ts=2020-06-26T20:31:42.784Z caller=main.go:339 host_details="(Linux 5.4.0-1012-raspi #12-Ubuntu SMP Wed May 27 04:08:35 UTC 2020 aarch64 6197a2742451 (none))"
panic: syscall.Getrlimit failed: operation not permitted

goroutine 1 [running]:
github.com/prometheus/prometheus/pkg/runtime.getLimits(0x7, 0x0, 0x0, 0x2, 0x2)
        /app/pkg/runtime/limits_default.go:39 +0x2b4
github.com/prometheus/prometheus/pkg/runtime.FdLimits(...)
        /app/pkg/runtime/limits_default.go:46
main.main()
        /app/cmd/prometheus/main.go:340 +0x3b80

```

**Comment by jsrz** [comment]
Are there plans to release a new docker image with an updated golang?

**Comment by tianon** [comment]
For the curious, the root cause of this denial is a bug in `libseccomp` where it wrongly presumed `getrlimit` didn't exist on these ARM arches: https://github.com/seccomp/libseccomp/issues/234 (apparently fixed in `libseccomp` 2.4.4).

For a workaround that doesn't involve the not-recommended `--privileged` insecurity hammer, `--security-opt seccomp:unconfined` should do the trick.

**Comment by violentmagician** [comment]
it works. i will now be able to deploy this on my pi4s
